x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/allocation/TrainedModelAllocationClusterService.java

```diff
@@ -448,8 +439,24 @@ static boolean shouldAllocateModels(final ClusterChangedEvent event) {
 
     Optional<String> nodeHasCapacity(ClusterState state, StartTrainedModelDeploymentAction.TaskParams params, DiscoveryNode node) {
         NodeLoad load = nodeLoadDetector.detectNodeLoad(state, true, node, Integer.MAX_VALUE, maxMemoryPercentage, useAuto);
+        return handleNodeLoad(load, node.getId(), params);
+    }
+
+    Optional<String> nodeHasCapacity(
+        ClusterState state,
+        TrainedModelAllocationMetadata.Builder builder,
+        StartTrainedModelDeploymentAction.TaskParams params,
+        DiscoveryNode node
+    ) {
+        NodeLoad load = builder.isChanged()
```

Comment:
This behaviour might be surprising or cause a subtle bug as if you pass the builder (this is an overloaded method) you expect it to be used. I can't think of reason not to call the `builder.build()` did you have something in mind?

Reply:
>  I can't think of reason not to call the builder.build() did you have something in mind?

It just seems unnecessary if nothing changed. The object could be pretty big and constructing a brand new object (deeply copying all the hashtables), could be expensive in the case when a node is simply leaving the cluster

Reply:
It works fine in this case because we know it is ok to fall back on the `TrainedModelAllocationMetadata` in `ClusterState state` but given a few refactors and a little bit of time that assumption may no longer hold and lead unexpected behaviour.

Perhaps it is not as neat but you could test `builder.isChanged()` outside this function and then call the correct overload or at least add a comment

Reply:
> if you pass the builder (this is an overloaded method) you expect it to be used

I suppose it is a sneaky switch. I will update it to be more clear so other callers of the method aren't surprised.